src/native/libs/System.Security.Cryptography.Native/pal_ssl.c

```diff
@@ -673,7 +673,14 @@ int CryptoNative_SslCtxSetCaching(SSL_CTX* ctx, int mode, int cacheSize, int con
     // void shim functions don't lead to exceptions, so skip the unconditional error clearing.
 
     // We never reuse same CTX for both client and server
-    SSL_CTX_ctrl(ctx, SSL_CTRL_SET_SESS_CACHE_MODE,  mode ? SSL_SESS_CACHE_BOTH : SSL_SESS_CACHE_OFF, NULL);
+    int modeFlags = SSL_SESS_CACHE_BOTH;
+    if (newSessionCb && removeSessionCb)
+    {
+        // sessions are completely controlled externally (from .NET), disable internal cache
```

Comment:
I don't think the is quite true. The existing code still assumes that we would be called when OpenSSL wants to remove entry and for example there is no implementation of upper limit. And we do not keep copy of the session.  So if this makes big difference in size we should dig deeper into it.

Reply:
I see, I think we should be manually removing the session from the internal cache instead:

> In TLSv1.3 it is recommended that each SSL_SESSION object is only used for
resumption once. One way of enforcing that is for applications to call
L<SSL_CTX_remove_session(3)> after a session has been used.

Reply:
Looks like OpenSSL will call RemoveSession on its own, and if we remove it prematurely it will break resume. So we may still use sessions more than once, that may be better left for future change.

Reply:
Do we know if that respects the 1x for each ticket? The report given to me claimed different memory use between 1.2 and 1.3 - all going to single server. But there were parallel requests and I'm not sure we have good test case for it. 

While the change looks reasonable to me I feel we should make sure we handle all the TLS 1.3 cases as best as we can. I'm wondering if we have multiple TLS session to the  same destination if we can get multiple tickets - each valid for one more resume. The current logic really maps to TLS 1.2 where we can simply keep the last session and use it later as many times as we want to.

Reply:
Let's continue the discussion in the main thread